lambda-buffers-compiler/src/LambdaBuffers/Compiler/TypeClass/Solve.hs

```diff
@@ -0,0 +1,145 @@
+
+{- Given two patterns (which are hopefully structurally similar), gather a list of all substitutions
+   from the PatVars in the first argument to the concrete types (hopefully!) in the second argument
+-}
+getSubs :: Pat -> Pat -> [(Text, Pat)] -- should be a set, whatever
+getSubs (VarP s) t = [(s, t)]
+getSubs (x :* xs) (x' :* xs') = getSubs x x' <> getSubs xs xs'
+getSubs (l := t) (l' := t') = getSubs l l' <> getSubs t t'
+getSubs (ProdP xs) (ProdP xs') = getSubs xs xs'
+getSubs (RecP xs) (RecP xs') = getSubs xs xs'
+getSubs (SumP xs) (SumP xs') = getSubs xs xs'
+getSubs (AppP t1 t2) (AppP t1' t2') = getSubs t1 t1' <> getSubs t2 t2'
+getSubs (RefP n t) (RefP n' t') = getSubs n n' <> getSubs t t'
+getSubs (DecP a b c) (DecP a' b' c') = getSubs a a' <> getSubs b b' <> getSubs c c'
+getSubs _ _ = []
+
+-- should be vastly more efficient than Data.List.Nub
+deduplicate :: Ord a => [a] -> [a]
+deduplicate = S.toList . S.fromList
+
+-- is the first pattern a substitution instance of the second
+isSubstitutionOf :: Pat -> Pat -> Bool
+isSubstitutionOf p1 p2 = matches p2 p1
+
+compareSpecificity :: Pat -> Pat -> Ordering
+compareSpecificity p1 p2
+  | p1
+      `isSubstitutionOf` p2
+      && p2
+      `isSubstitutionOf` p1 =
+      EQ
+  | p1 `isSubstitutionOf` p2 = LT
+  | otherwise = GT
+
+sortOnSpecificity :: Pat -> Class -> [Rule] -> [Rule]
+sortOnSpecificity p c ps =
+  sortBy (\a1 a2 -> compareSpecificity (ruleHeadPat a1) (ruleHeadPat a2)) $
+    filter matchPatAndClass ps
+  where
+    matchPatAndClass :: Rule -> Bool
+    matchPatAndClass r =
+      ruleHeadClass r == c
+        && ruleHeadPat r
+        `matches` p
+
+mostSpecificInstance :: Pat -> Class -> [Rule] -> Maybe Rule
+mostSpecificInstance p c ps = case sortOnSpecificity p c ps of
+  [] -> Nothing
+  (x : _) -> Just x
+
+{- Given a list of instances (the initial scope), determines whether we can derive
+   an instance of the Class argument for the Pat argument. A result of [] indicates that there are
+   no remaining subgoals and that the constraint has been solved.
+   NOTE: At the moment this handles superclasses differently than you might expect -
+         instead of assuming that the superclasses for all in-scope classes are defined,
+         we check that those constraints can be solved before affirmatively judging that the
+         target constraint has been solved. I *think* that makes sense in this context (whereas in Haskell
+         it doesn't b/c it's *impossible* to have `instance Foo X` if the definition of Foo is
+         `class Bar y => Foo y` without an `instance Bar X`)
+-}
+solve ::
+  [Rule] -> -- all instance rules in scope. WE ASSUME THESE HAVE ALREADY BEEN GENERATED, SOMEWHERE
+  Constraint -> -- constraint we're trying to solve
+  [Constraint] -- subgoals that cannot be solved for w/ the current rule set
+solve inScope cst@(C c pat) =
+  -- First, we look for the most specific instance...
+  case mostSpecificInstance pat c inScope of
```

Comment:
I need some clarity on this, you should be needing this because we don't have duplicate instances right?

Reply:
Honestly I was doing it because that's what the research I did lead me to believe I should do. 

The problem is really overlapping instances, not duplicates (the machinery that prepares this uses `Set Rule` so there's not even a point in worrying about duplicates here). 

After thinking about this, though, you're right that `mostSpecificInstance` doesn't really make much sense. If there are multiple matching instances then we have overlapping instances which we can't have. I'll revert that to something like the old implementation today, which threw an error if multiple in-scope instances match a given head. 

(There might be a place for overlapping instances if we wanted to express more sophisticated structural rules, and we may need them in a limited capacity to do things with `Nat`s if we ever implement those)

Reply:
I fixed this in the commits i'm about to push